app/artifact-cas/internal/service/bytestream.go

```diff
@@ -161,6 +164,11 @@ func (s *ByteStreamService) Read(req *bytestream.ReadRequest, stream bytestream.
 		return sl.LogAndMaskErr(err, s.log)
 	}
 
+	// check if the file has been tampered with and notify the client
+	if sw.GetChecksum() != req.ResourceName {
```

Comment:
This is a verification server-side that the file has not been tampered with and we deliver only the file if it matches the digest. Note that we already do this check client-side but now server side makes it more robust.